### ISSUE TITLE
Feedback on private imports

### ISSUE THREAD

**Issue by bwilkerson** [dart-lang/language#1809]
The [Packaged Libraries](https://github.com/dart-lang/language/blob/master/working/packaged-libraries/feature-specification.md) proposal includes a section on private imports and includes the following questions / comments.

> __TODO: Better syntax? Allow only importing some private names?__

It would be inconsistent if the language didn't support hiding private names because hiding is one of the primary ways users can deal with conflicting imports. (How common it is in practice, compared to using prefixes, would be an interesting thing to measure.) If we're going to allow hiding individual top-level names then we should be able to show them as well.

If we need to allow individual private names in the `show` clause, then we might want a different syntax for the "show all" case because `_` is a valid private identifier and would hence be ambiguous.

Another possibility would be to _only_ allow showing individual names on the grounds that it's safer than importing everything. It would be interesting to know how many names would need to be shown if they had to be listed explicitly.

> __TODO: We could consider something like #1627 to allow importing only certain private instance members.__

If we were to support hiding and showing instance members for other reasons, then consistency would suggest that we support private members as well. Without it there's a chance that introducing a private method in one library might cause a pre-existing method in a different library to override it, either with or without error. If this could happen across packages I'd say that hiding the unexpectedly overridden member was fairly important, but within a single package it should be fairly easy to rename one of the members.

**Comment by bernaferrari** [comment]
Random question, but, at this point, isn't it simpler to just add `open`/`private`/`interal`/`public`/`protected`/something like this? `_` is supposed to be private, now the proposal includes a way to access what is private, but you are proposing a way to hide access to what is private (if I understood correctly), so you want a double private mechanism.

**Comment by munificent** [comment]
> The [Packaged Libraries](https://github.com/dart-lang/language/blob/master/working/packaged-libraries/feature-specification.md) proposal includes a section on private imports and includes the following questions / comments.
> 
> > **TODO: Better syntax? Allow only importing some private names?**
> 
> It would be inconsistent if the language didn't support hiding private names because hiding is one of the primary ways users can deal with conflicting imports.

This is true. But as an affordance, that capability is a *lot* more important across packages than within a single package. Hiding is useful because if some library *you don't control* adds a name that collides, you have a remedy.

But with importing private names, the names are always under your control and you always have the luxury of renaming them if you want (since they are private and thus renaming them won't break any external users).

> If we're going to allow hiding individual top-level names then we should be able to show them as well.

I generally agree though that it would be nice to be able to hide and show private imported names too.

> Another possibility would be to _only_ allow showing individual names on the grounds that it's safer than importing everything. It would be interesting to know how many names would need to be shown if they had to be listed explicitly.

I considered that, but this feature also gives you access to private *instance members* too. So it's not just about top-level names, which is all `show` and `hide` are currently concerned with. So if we did this, we'd need to come up with some way to import private instances names too.

My impression is that `show` and `hide` are pretty rare. (I looked once years ago, but don't remember the number. I think it was something like roughly 90% of imports didn't use a prefix, show, or hide.) Given that, I think the right default here is probably just to import all private names. That's terse, which is nice.

**Comment by munificent** [comment]
> Random question, but, at this point, isn't it simpler to just add `open`/`private`/`internal`/`public`/`protected`/something like this?

The proposal discusses all of that too.

> `_` is supposed to be private, now the proposal includes a way to access what is private, but you are proposing a way to hide access to what is private (if I understood correctly), so you want a double private mechanism. 

No, there's no need for "double private". Private today means both "inaccessible to libraries outside of this package" and "inaccessible to other libraries in the same package". The private imports proposal gives you a way to express the former without the latter.
